sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/Node.java

```diff
@@ -0,0 +1,520 @@
+
+    private Node(UUID parentId, Type type, String role, Method method, EnumSet<Bypass> subNodeExecutionModes) {
+        this.type = type;
+        this.role = role;
+        this.parentId = parentId;
+        this.subNodeExecutionModes = subNodeExecutionModes;
+        this.method = method;
+        this.phase = Phase.CREATED;
+        this.condition = null;
+    }
+
+    private Node(Builder builder, UUID parentId, Type type) {
+        this.type = type;
+        this.parentId = parentId;
+        this.role = builder.role;
+        this.name = builder.name;
+        this.description = builder.description;
+        this.historyId = builder.historyId;
+        this.subNodeExecutionModes = builder.subNodeExecutionModes;
+        this.condition = builder.condition;
+        this.method = builder.method;
+        this.parameters.putAll(builder.parameters);
+        this.attributes.putAll(builder.attributes);
+    }
+
+    public static Node createRootNode(String role, EnumSet<Bypass> subNodesSkippingFlags) {
+        return new Node(UUID.randomUUID(), Type.HOLDER, role, null, subNodesSkippingFlags);
+    }
+
+    public UUID getRuntimeId() {
+        return this.runtimeId;
+    }
+
+    public UUID getParentId() {
+        return this.parentId;
+    }
+
+    public String getHistoryId() {
+        return this.historyId;
+    }
+
+    public String getDescription() {
+        return this.description;
+    }
+
+    public Optional<Object> getAttribute(String name) {
+        return Optional.ofNullable(this.attributes.get(name));
+    }
+
+    public Stream<Map.Entry<String, Object>> attributesStream() {
+        return this.attributes.entrySet().stream();
+    }
+
+    public String getRole() {
+        return this.role;
+    }
+
+    public String getCurrentState() {
+        return "[ Status: '" + status + "'; Phase: '" + phase + "'; Is bypassed: '" + isBypassed + "'; ]";
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public Method getMethod() {
+        return method;
+    }
+
+    public Map<String, Object> getMethodParameters() {
+        return new LinkedHashMap<>(parameters);
+    }
+
+    public Optional<Throwable> getThrowable() {
+        return Optional.ofNullable(throwable);
+    }
+
+    public Optional<Condition> getCondition() {
+        return Optional.ofNullable(this.condition);
+    }
+
+    public boolean isReadyForInvoke() {
+        return Type.INVOKABLE.equals(type) && !isBypassed && Status.STARTED.equals(status) && Phase.EXECUTING.equals(phase);
+    }
+
+    public void completePreparation() {
+        this.phase = Phase.PREPARED;
+    }
+
+    public void scheduleExecution() {
+        this.phase = Phase.SCHEDULED;
+    }
+
+    public void bypassExecution() {
+        this.isBypassed = true;
+    }
+
+    public boolean isBypassed() {
+        return this.isBypassed;
+    }
+
+    public boolean isEmptyHolder() {
+        return Type.HOLDER.equals(this.type) && this.children.size() == 0;
```

Comment:
Isn't it better to use some isEmpty() method?

Reply:
updated with suggestion